### ISSUE TITLE
3.9 Regression Merge Types

### ISSUE THREAD

**Issue by micheleangioni** [microsoft/TypeScript#38571]
**TypeScript Version:**  3.9.2

**Search Terms:** Regression, merge types

**Code**

```ts
type Merge<T, S> = Omit<T, Extract<keyof T, keyof S>> & S;

type WithNumberId = {
  _id: number,
  otherKey?: number,
};

type WithStringId = {
  _id: string,
  otherKey?: string,
};

function convertToStringId<T extends WithNumberId>(data: T): Merge<T, WithStringId> {
  const convertedData: Merge<T, WithStringId> = {
    ...data,
    _id: data._id.toString(),
  };

  if (data.otherKey) {
    convertedData.otherKey = data.otherKey.toString();
  }

  return convertedData;
}
```

**Expected behavior:** No error (3.8x is fine)

**Actual behavior:** Definition of `const convertedData` gives error

**Playground Link:** https://www.typescriptlang.org/play/?ssl=20&ssc=35&pln=20&pc=51#code/C4TwDgpgBAshBOBzCAeAKgGigZQHxQF4oB5AWwEth0sBRAD2HgEMBjKgawhAHsAzKTFE49+efADIcAbgBQM0JCgB1SgAsAcgFdSAIwQBJACaEoAbxlQoAfXKGAXFAB22vfAwWo3YKoQBpLgD8Ds66CO4AvrLy4NAq3tiM5I6IRibmljb2UADOicnull4+8P4gQTl5iBFRvJqObOTcjlAsTQBuCMBo3AnwSSmG6FAQDBCOhtnKalqh8Ea4ABSGTMBMDmgAlA5wSKiCcaq9-fNmHq2OuS3tnRCGACIra7AIyNRT8ZUnROmWUAB0AOWqwKv0yDiBTD+mT+wB6lQWGxBkTklnI-CWjz+RT8XA2p1+V0cHXgwFuD1WWO8OJAJghlOKpRhcL6yQRsks4RRUHgEGAmngzXOxNJ90esk5QA

**Related Issues:** Maybe https://github.com/microsoft/TypeScript/issues/38460

**Comment by x87** [comment]
This is expected due to a breaking change #37195

**Comment by micheleangioni** [comment]
Are you sure @x87 ? The **NON** optional __id_ property is causing the problem.

**Comment by x87** [comment]
@micheleangioni the error says:

```
Types of property 'otherKey' are incompatible.
      Type 'number | undefined' is not assignable to type 'string | undefined'.
        Type 'number' is not assignable to type 'string | undefined'.(2322)
```
if you remove the `_id` property from types `WithNumberId` and `WithNumberString` altogether the error remains.
if you remove the `otherKey` property it's gone.

**Comment by RyanCavanaugh** [comment]
This is the correct behavior. At its initialization, `convertedData` has a declared property of type `string` inhabited (possibly) by a value of type `number`.

**Comment by micheleangioni** [comment]
Ok, thanks @RyanCavanaugh for pointing out is was due to 'otherKey' :)
